File: regular_assignments/day_009/sum_of_all_subarrays/solution.rb
class Solution
  def call
    subarraySum([1, 2, 3]) # 20
    subarraySum([2, 1, 3]) # 19
  end

  def subarraySum(arr)
    sum = 0
    i = 0
    while i < arr.length
      freq = (i + 1) * (arr.length - i)
      contribution = freq * arr[i]
      sum += contribution
      i += 1
    end
    sum
  end
end

# puts '-' * 100
# puts Solution.new.subarraySum([1,2,3])
# puts '-' * 100
# puts Solution.new.subarraySum([2,1,3])

File: regular_assignments/day_009/sum_of_all_subarrays/solution.go
package sumOfAllSubArrays

func Call() {
	subArraySum([]int{1, 2, 3})
	subArraySum([]int{2, 1, 3})
}

func subArraySum(arr []int) int64 {
	sum := 0
	for i := 0; i < len(arr); i++ {
		freq := (i + 1) * (len(arr) - i)
		contribution := freq * arr[i]
		sum += contribution
	}
	return int64(sum)
}
